contracts/src/rewards/RewardsManager.sol

```diff
@@ -0,0 +1,122 @@
+        initialized = true;
+    }
+
+    function setInflationRate(SD59x18 _inflationRate) external {
+        maxInflation = _inflationRate;
+    }
+
+    function setMinInflationFactor(SD59x18 _minInflationFactor) external {
+        minInflationFactor = _minInflationFactor;
+    }
+
+    function setPrice(SD59x18 _price) external {
+        price = _price;
+    }
+
+    function setCost(SD59x18 _cost) external {
+        cost = _cost;
+    }
+
+    /**
+     * @dev Distribute reward tokens to addresses.
+     * @param epoch epoch for which the rewards to be distributed
+     */
+    function distributeRewards(uint256 epoch) public onlyController {
+        require(epochManager.epoch() != epoch, "cannot distribute rewards for current epoch");
+        require(!rewardsDistribution[epoch], "rewards already distributed for this epoch");
+
+        SD59x18 _uActual = convert(int256(rewardsAggregator.getDataForEpoch(epoch)));
+        SD59x18 _uPotential = convert(int256(rewardsAggregator.getAvgUsageNEpochs(epoch)));
+
+        SD59x18 _totalMint = _getMintRate(_uActual, _uPotential);
+        // todo: variable distribution
+        // 75% goes to edge node
+        SD59x18 _toEdgeNode = _totalMint.mul(sd(0.75e18));
+        string[] memory publicKeys = rewardsAggregator.getPublicKeys(epoch);
+
+        for (uint256 i = 0; i < publicKeys.length; i++) {
+            uint256 dataServedByNode = rewardsAggregator.getDataServedByNode(publicKeys[i], epoch);
+            SD59x18 servedPercentage = convert(int256(dataServedByNode)).div(_uActual);
+            SD59x18 rewardsAmount = servedPercentage.mul(_toEdgeNode);
+            // check if the node with public key is white listed
+            (address _to,,,,) = nodeRegistry.whitelist(publicKeys[i]);
+            fleekToken.mint(_to, intoUint256(rewardsAmount));
+            emit RewardMinted(_to, intoUint256(rewardsAmount));
+        }
+        rewardsDistribution[epoch] = true;
+    }
+
+    /**
+     * @dev calculate the minting based actual usage and potential usage
+     * @param _uActual actual usage in the epoch for which the minting is calculated
+     * @param _uPotential potential usage in the epoch for which the minting is calculated
+     */
+    function _getMintRate(SD59x18 _uActual, SD59x18 _uPotential) private returns (SD59x18 totalMint) {
+        // Equation 2 from the paper
+        // delta U = (_uActual - _uPotential)/uPotential
+        SD59x18 _deltaUNumerator = _uActual.sub(_uPotential);
+        SD59x18 _deltaU = _deltaUNumerator.div(_uPotential);
+
+        // Equation 3 from the paper
+        SD59x18 potentialFactor = UNIT.sub(((price).div(cost)).mul(_deltaU));
+        SD59x18 firstMax = sd(MathUtils.signedMax(intoInt256(potentialFactor), intoInt256(minInflationFactor)));
+        SD59x18 dynamicInflation = firstMax.mul(inflationInLastEpoch);
+        SD59x18 currentInflation = sd(MathUtils.signedMin(intoInt256(dynamicInflation), intoInt256(maxInflation)));
```

Comment:
wasn't `minInflationFactor` factor added to prevent that? and I think in your simulation 0.5 is `minInflationFactor`. According to the equation in the paper, this is how `minInflationFactor` is used
 $ i_{current} = min(i_{prev} * (max(1 - \frac {P_{usd}}{C_{avg}} * \Delta U, \space minInflationFactor) ), \space i_{max}) $
 
 did we change some thing after this

Reply:
> and I think in your simulation 0.5 is minInflationFactor

Yeah, you are right, it's `currentInflation = min(currentInflation, minInflationFactor * maxInflation)`.  

> wasn't minInflationFactor factor added to prevent that?

Yes, but the simulation spits out different results depending on whether `currentInflation = min(currentInflation, minInflationFactor * maxInflation)` is included or not.

Reply:
thanks for pointing out. I was forgetting this part that we added to the simulations.
changing the overall equation to bring in this change

$/ i_{current} =
    \left\{ 
  \begin{array}{ c l }
    max(i_{prev} * \Delta U * (1 - \frac {P_{usd}}{C_{avg}}), \space minInflationFactor * i_{max})  & \quad \textrm{if } \Delta U \geq 0 \\
    min(i_{prev} * \Delta U * (1 - \frac {P_{usd}}{C_{avg}}), \space i_{max})                 & \quad \textrm{otherwise}
  \end{array}
\right.